src/core/stdcpp/array.d

```diff
@@ -0,0 +1,149 @@
+/**
+ * D header file for interaction with C++ std::array.
+ *
+ * Copyright: Copyright (c) 2018 D Language Foundation
+ * License: Distributed under the
+ *      $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors:   Manu Evans
+ * Source:    $(DRUNTIMESRC core/stdcpp/array.d)
+ */
+
+module core.stdcpp.array;
+
+// hacks to support DMD on Win32
+version (CppRuntime_Microsoft)
+{
+    version = CppRuntime_Windows; // use the MS runtime ABI for win32
+}
+else version (CppRuntime_DigitalMars)
+{
+    version = CppRuntime_Windows; // use the MS runtime ABI for win32
+    pragma(msg, "std::array not supported by DMC");
+}
+
+extern(C++, "std"):
+
+/**
+ * D language counterpart to C++ std::array.
+ *
+ * C++ reference: $(LINK2 https://en.cppreference.com/w/cpp/container/array)
+ */
+extern(C++, class) struct array(T, size_t N)
+{
+extern(D):
+pragma(inline, true):
+
+    ///
+    alias size_type = size_t;
+    ///
+    alias difference_type = ptrdiff_t;
+    ///
+    alias value_type = T;
+    ///
+    alias pointer = T*;
+    ///
+    alias const_pointer = const(T)*;
+
+    ///
+    alias as_array this;
+
+    /// Variadic constructor
+    this(T[N] args ...)                                             { this[] = args[]; }
+
+    ///
+    size_type size() const nothrow @safe @nogc                      { return N; }
+    ///
+    alias length = size;
+    ///
+    alias opDollar = length;
+    ///
+    size_type max_size() const nothrow @safe @nogc                  { return N; }
+    ///
+    bool empty() const nothrow @safe @nogc                          { return N == 0; }
+
+    ///
+    ref inout(T) front() inout nothrow @safe @nogc                  { static if (N > 0) { return this[0]; } else { return as_array()[][0]; /* HACK: force OOB */ } }
+    ///
+    ref inout(T) back() inout nothrow @safe @nogc                   { static if (N > 0) { return this[N-1]; } else { return as_array()[][0]; /* HACK: force OOB */ } }
+
+    ///
+    void fill()(auto ref const(T) value) @safe @nogc                { this[] = value; }
```

Comment:
From what I can see (if you follow my suggestion about `RangeError` you can just slap `nothrow @nogc @safe:` at the top of this `struct`.

Reply:
Actually, thinking on this just now; this function is neither `@safe` nor `@nogc`; T's assignment doesn't promise those properties...
I guess I should remove them here and allow it to infer from the assignments.